lib/trace.js

```diff
@@ -1,24 +1,121 @@
 "use strict";
 
-const { exists } = require("./util");
+const fs = require("fs");
+const path = require("path");
+const { promisify } = require("util");
+
+const { parse } = require("acorn-node");
+const { getDeps } = require("./extract");
+
+// Helpers
+const resolve = promisify(require("resolve"));
+const readFile = promisify(fs.readFile);
+const uniq = (items) => Array.from(new Set(items)).sort();
+
+// HELPER: Recursively track and trace dependency paths.
+// - `srcPaths`: If provided, **don't** add to resolvedDepPaths
+// - `depPaths`: If provided, **do** add to resolvedDepPaths
+const _recurseDeps = async ({ srcPaths, depPaths = [], ignores, tracedDepPaths }) => {
+  // Start **only** with depPaths.
+  let resolvedDepPaths = Array.from(depPaths);
+
+  // If srcPaths is provided, now **replace** depPaths.
+  // (Allows easier accomodation of `traceFile` vs. `traceFiles`)
+  depPaths = srcPaths || depPaths;
+
+  // TODO(6): Consider parallelizing file traversal
+  // https://github.com/FormidableLabs/trace-deps/issues/6
+  for (const depPath of depPaths) {
+    if (!tracedDepPaths.has(depPath) && (/\.(js|mjs)$/).test(depPath)) {
+      // Mark encountered to avoid future recursion.
+      tracedDepPaths.add(depPath);
+
+      // Recurse.
+      // eslint-disable-next-line no-use-before-define
+      const recursedDepPaths = await traceFile({ srcPath: depPath, ignores, tracedDepPaths });
+
+      // Aggregate.
+      resolvedDepPaths = resolvedDepPaths.concat(recursedDepPaths);
+    }
+  }
+
+  return resolvedDepPaths;
+};
 
 /**
  * Trace and return on-disk locations of all file dependencies from a source file.
  *
- * @param {*}       opts              options object
- * @param {string}  opts.file         source file to trace
- * @returns {Promise<Array<String>>}  list of relative paths to on-disk dependencies
+ * **Note**: An internal set, `tracedDepPaths`, tracks every file we encounter
```

Comment:
@bmathews -- Added this note as an internal code comment. Could do externally, but I'm guessing this might be enough for now?

Reply:
:+1: that's great.